ntoskrnl/ke/amd64/usercall.c

```diff
@@ -253,8 +253,11 @@ KeUserModeCallback(
     /* Enter a SEH Block */
     _SEH2_TRY
     {
-        /* Calculate and align the stack size */
-        UserArguments = (PUCHAR)ALIGN_DOWN_POINTER_BY(OldStack - ArgumentLength, sizeof(PVOID));
+        /* Calculate and align the stack. This is unaligned by 8 bytes, since the following
+           UCALLOUT_FRAME compensates for that and on entry we already have a full stack
+           frame with home space for the next call, i.e. we are already inside the function
+           body and the stack needs to be 16 byte aligned. */
+        UserArguments = (PUCHAR)ALIGN_DOWN_POINTER_BY(OldStack - ArgumentLength, 16) - 8;
```

Comment:
Maybe we should have UCALLOUT_FRAME have a 16-bytes aligned size.

Reply:
It's based on information in ksamd64.inc from the Windows SDK:
```
;
;  Usermode Callout User Frame Definitions
;

CkBuffer equ 00020H
CkLength equ 00028H
CkApiNumber equ 0002CH
CkRsp equ 00048H
CkRip equ 00030H
CalloutFrameLength equ 00058H
```
This matches this result (from reactos x64):
```
kd> dt -r _UCALLOUT_FRAME
nt!_UCALLOUT_FRAME
   +0x000 P1Home           : Uint8B
   +0x008 P2Home           : Uint8B
   +0x010 P3Home           : Uint8B
   +0x018 P4Home           : Uint8B
   +0x020 Buffer           : Ptr64 Void
   +0x028 Length           : Uint4B
   +0x02c ApiNumber        : Uint4B
   +0x030 MachineFrame     : _MACHINE_FRAME
      +0x000 Rip              : Uint8B
      +0x008 SegCs            : Uint2B
      +0x00a Fill1            : [3] Uint2B
      +0x010 EFlags           : Uint4B
      +0x014 Fill2            : Uint4B
      +0x018 Rsp              : Uint8B
      +0x020 SegSs            : Uint2B
      +0x022 Fill3            : [3] Uint2B

```

Reply:
Alright then.